WordPress/Sniffs/WP/PostTypeSniff.php

```diff
@@ -0,0 +1,117 @@
+ * @link    https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WordPressCS\WordPress\Sniffs\WP;
+
+use WordPressCS\WordPress\AbstractFunctionParameterSniff;
+
+/**
+ * Validates post type names.
+ *
+ * @package WPCS\WordPressCodingStandards
+ *
+ * @since X.X.X
+ */
+class PostTypeSniff extends AbstractFunctionParameterSniff {
+
+	const MAX_POST_TYPE_LENGTH = 20;
+
+	/**
+	 * Array of functions that must be checked.
+	 *
+	 * @since X.X.X
+	 *
+	 * @var array Multidimensional array with parameter details.
+	 *     $target_functions = array(
+	 *         (string) Function name.
+	 *     );
+	 */
+	protected $target_functions = array(
+		'register_post_type'   => true,
+	);
+
+	/**
+	 * Array of reserved post type names which can not be used by themes and plugins.
+	 *
+	 * @since X.X.X
+	 *
+	 * @var array
+	 */
+	protected $reserved_names = array(
+		'post'                => true,
+		'page'                => true,
+		'attachment'          => true,
+		'revision'            => true,
+		'nav_menu_item'       => true,
+		'custom_css'          => true,
+		'customize_changeset' => true,
+		'oembed_cache'        => true,
+		'user_request'        => true,
+		'wp_block'            => true,
+		'action'              => true,
+		'author'              => true,
+		'order'               => true,
+		'theme'               => true,
+	);
+
+	/**
+	 * Process the parameter of a matched function.
+	 *
+	 * Errors on invalid post type names when reserved keywords are used,
+	 * the post type is too long, or contains invalid characters.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param int    $stackPtr        The position of the current token in the stack.
+	 * @param array  $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 * @param array  $parameters      Array with information about the parameters.
+	 *
+	 * @return void
+	 */
+	public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
+		$post_type = $this->strip_quotes( $parameters[1]['raw'] );
+		if ( isset( $this->reserved_names[ $post_type ] ) ) {
+			$this->phpcsFile->addError(
+				'%s() called with reserved post type %s. Reserved post types should not be used as they interfere with other WordPress functions.',
+				$stackPtr,
+				$this->string_to_errorcode( $post_type . '_Reserved' ),
+				array(
+					$matched_content,
+					$parameters[1]['raw'],
+				)
+			);
+		}
+
+		if ( strlen( $post_type ) > self::MAX_POST_TYPE_LENGTH ) {
+			$this->phpcsFile->addError(
+				'%s() called with invalid post type %s. Post type longer than %d characters.',
+				$stackPtr,
+				$this->string_to_errorcode( $post_type . '_TooLong' ),
+				array(
+					$matched_content,
+					$parameters[1]['raw'],
+					self::MAX_POST_TYPE_LENGTH,
+				)
+			);
+		}
+
+		if ( preg_match( '/[A-Z _]/', $post_type ) === 1 ) {
```

Comment:
Are emoji or other Unicode characters (think umlauts in German or something) allowed? Right now, you're _blacklisting_ uppercase, space and underscore; a _whitelisting_ of what is allowed might be easier to manage.

Reply:
The blacklist comes from this comment 

> max. 20 characters, cannot contain capital letters, underscores or spaces
found here https://codex.wordpress.org/Function_Reference/register_post_type#Parameters

I don't see why using unicode characters would be wrong, as they can correctly be saved to the database. They could possibly cause issues with template loading. The whole idea of using non-ascii characters as a post type is probably ill advised. 

I also have no idea why underscores would be disallowed apart from the comment in the documentation. 

Maybe we can add a whitelist, and post types with non-whitelisted characters trigger a warning/risky notice?

Reply:
I think the documentation was written from an American English point of view and didn't even consider that people would want to use non-ascii characters.

The post type name is used/can be used in all sorts of places, including query variables, theme templates file names, hook names etc. I imagine there may even be callbacks which use the post type name as the PHP/JS function name.
I dread to start researching this in depth as I'm very afraid of what we would encounter....

With that in mind, I have a feeling that the _real_ rule is "_only use lowercase a-z and possibly a `-` dash_".

> Maybe we can add a whitelist, and post types with non-whitelisted characters trigger a warning/risky notice?

Sounds like a very good idea :+1: , a stern warning at that if it were up me.
This would need additional unit tests, both with unicode post type names, as well as some simple ascii names which I think would be problematic anyway, even though based on the existing documentation they would be allowed...
Think: `q&a` (query vars), `questions?` (query vars), `my/post/type` (template file names), `prefix\post-type`, `comma,separated,list`

Reply:
Just looked at the plugin handbook;

> To ensure forward compatibility, do not use wp_ as your identifier — it is being used by WordPress core.
>
> Make sure your custom post type identifier does not exceed 20 characters as the post_type column in the database is currently a VARCHAR field of that length.
> 
> If your identifier is too generic — for example: product. It may conflict with other plugins or themes.

That blacklists one more name (or is it a prefix ?) as reserved + explains the why of the hard 20 char limit.
Also: as the database charset is configurable in `wp-config.php` using the `DB_CHARSET` constant, we should presume that not everyone uses `utf8`.

Reply:
Hmm... just noticed that the example in the plugin handbook uses an `_` underscore in the post type name.... _sigh_

Reply:
Related: https://wordpress.org/support/article/post-types/

Reply:
Just a though - I wonder if the "_no underscores_" rule is a documentation error.

Even just looking at the WP native post types, underscores are used.
All Custom Post Type generators I've seen also support them (and I would have expected them to have received bug reports if the code didn't work).

Reply:
Confirmed - underscores are allowed and only alphanumeric characters are supported:

> Must not exceed 20 characters and may only contain lowercase alphanumeric characters, dashes, and underscores.

Source: https://developer.wordpress.org/reference/functions/register_post_type/